vm/devices/storage/disk_nvme/nvme_driver/src/queue_pair.rs

```diff
@@ -461,6 +553,56 @@ impl QueueHandler {
             }
         }
     }
+
+    /// Save queue data for servicing.
+    pub async fn save(&self) -> anyhow::Result<QueuePairSavedState> {
+        let mut pending_cmds: Vec<PendingCommandSavedState> = Vec::new();
+        for cmd in &self.commands {
+            let mut command: [u8; 64] = [0; 64];
+            command.copy_from_slice(cmd.1.command.as_bytes());
+            let command = PendingCommandSavedState {
+                command,
+                cid: cmd.0 as u16,
+            };
+            pending_cmds.push(command);
+        }
+        // The data is collected from both QueuePair and QueueHandler.
+        Ok(QueuePairSavedState {
+            max_cids: self.max_cids,
+            sq_state: self.sq.save(),
+            cq_state: self.cq.save(),
+            pending_cmds,
+            cpu: 0,       // Will be updated by the caller.
+            msix: 0,      // Will be updated by the caller.
+            sq_addr: 0,   // Will be updated by the caller.
+            cq_addr: 0,   // Will be updated by the caller.
+            base_va: 0,   // Will be updated by the caller.
+            mem_len: 0,   // Will be updated by the caller.
+            pfns: vec![], // Will be updated by the caller.
+        })
+    }
+
+    /// Restore queue data after servicing.
+    pub fn restore(&mut self, saved_state: &QueuePairSavedState) -> anyhow::Result<()> {
```

Comment:
A key thing to reason about is what happens to requests that were in flight at save. It seems that the design is that we will keep those requests in flight (i.e., we won't wait for them to complete before saving/after restoring), right? And so storvsp's existing restore path will reissue any IOs that were in flight, but they won't wait for old IOs.

I think this is a problem, because new IOs can race with the original IO and cause data loss. Here's an example of what can go wrong:

1. Before servicing, the guest writes "A" to block 0x1000 with CID 1.
2. Before CID 1 completes, a servicing operation starts. CID 1 makes it into the saved state.
3. At restore, storvsp reissues the IO, so it writes "A" to block 0x1000 with CID 2.
3. Due to reordering in the backend, CID 2 completes before CID 1.
4. Storvsp completes the original guest IO.
5. The guest issues another IO, this time writing "B" to the same block 0x1000. Our driver issues this with CID 3.
6. CID 3 completes before CID 1.
7. Finally, some backend code gets around to processing CID 1. It writes "A" to block 0x1000.

There are variants on this, where the guest changes the memory that CID 1 was using and ASAP DMAs it late, after the IO was completed to the guest. This could even affects reads by corrupting guest memory.

The most reasonable solution, in my opinion, is to avoid issuing any more commands to the SQ until all IOs that were in flight at save time have completed. This should be simple to implement and to reason about. A more complicated solution would be to reason about what the in flight commands are doing and only block IOs that alias those pages. I don't think that's necessary for v1.

Reply:
Did I miss somewhere where this is handled?

Reply:
>> The most reasonable solution, in my opinion, is to avoid issuing any more commands to the SQ until all IOs that were in flight at save time have completed. This should be simple to implement and to reason about.

I think we cut off receiving mesh commands once servicing save request is received, let me double check that and at which point this is done.

Regarding draining of the outstanding (in-flight) I/Os - eventually we want to bypass this but for v1 it makes sense to drain them, for simplicity. Let me confirm the path. The draining happens for non-keepalive path.

Reply:
I think two things are important in the conversation above:

1. It's okay to drain IOs before issuing new IOs. But: unlike the non-keepalive path, the overall servicing operation should not block on outstanding IOs. This means that the replay will need to remain asynchronous.
2. I agree with John's analysis. It is unsafe to complete any IOs back to the guest until all outstanding IOs (IOs in progress at the save time to the underlying nvme device) complete from the NVMe device to the HCL environment. I think it is fine to start replaying the saved IOs while the in flight IOs are still in progress. This means you have two choices: (a) wait to begin replaying IOs until everything in flight completes, or (b) build some logic in the storvsp layer to not notify the guest until all the in flight IOs complete. (a) seems simpler to me.

In future changes, I think we will need to consider some sort of filtering (e.g. hold new IOs from guest that overlap LBA ranges, as John suggested in an offline conversation).

Reply:
So the in-flight IOs will be executed twice if my understanding is correct?

Reply:
We are saving Pending Commands slab so we can correctly find a CID for outstanding responses once we're back from blackout. We do not resubmit those commands.

Reply:
I mean in these steps "Before CID 1 completes, a servicing operation starts. CID 1 makes it into the saved state.
At restore, storvsp reissues the IO, so it writes "A" to block 0x1000 with CID 2." the same IO was issued twice with CID 1 and CID 2. Is it still true?

Reply:
That is true, I can observe that in the instrumented FW that same command may be issued twice - before and after servicing. Let me double check if CID is still same, as there is recent change to randomize CID in NVMe queue.